src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,432 @@
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    class MultipartPipeReader
+    {
+
+        public const int DefaultHeadersCountLimit = 16;
+        public const int DefaultHeadersLengthLimit = 1024 * 16;
+        private const int DefaultBufferSize = 1024 * 4;
+        private const int StackAllocThreshold = 128;
+
+        private readonly PipeReader _pipeReader;
+        private ReadOnlySequence<byte> _buffer;
+        private ReadResult _readResult = default;
+        private readonly BufferedReadStream _stream;
+        private readonly MultipartBoundary _boundary;
+        private MultipartReaderStream _currentStream;
+
+
+        public MultipartPipeReader(string boundary, PipeReader pipeReader)
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+            _buffer = default;
+
+            _stream = new BufferedReadStream(pipeReader.AsStream(), DefaultBufferSize);
+            _currentStream = new MultipartReaderStream(_stream, _boundary) { LengthLimit = HeadersLengthLimit };
+
+        }
+
+        /// <summary>
+        /// The limit for the number of headers to read.
+        /// </summary>
+        public int HeadersCountLimit { get; set; } = DefaultHeadersCountLimit;
+
+        /// <summary>
+        /// The combined size limit for headers per multipart section.
+        /// </summary>
+        public int HeadersLengthLimit { get; set; } = DefaultHeadersLengthLimit;
+
+        /// <summary>
+        /// The optional limit for the total response body length.
+        /// </summary>
+        public long? BodyLengthLimit { get; set; }
+
+
+        public async Task<MultipartSection> ReadNextSectionAsync(CancellationToken cancellationToken = new CancellationToken())
+        {
+
+            if (_buffer.IsEmpty)
+            {
+                _readResult = await _pipeReader.ReadAsync(cancellationToken);
+                _buffer = _readResult.Buffer;
+            }
+
+            SkipMetaData(ref _buffer, _readResult.IsCompleted);
+            if(_buffer.IsEmpty)
+            {
+                return null;
+            }
+            var headersAccumulator = new KeyValueAccumulator();
+            ParseHeaders(ref _buffer, ref headersAccumulator, _readResult.IsCompleted);
+            _boundary.ExpectLeadingCrlf = true;
+            var body = ReadBody(ref _buffer, _readResult.IsCompleted);
+            return new MultipartSection() { Headers = headersAccumulator.GetResults(), Body = new MemoryStream(body), BaseStreamOffset = 0 };
```

Comment:
This assumes the entire headers and body are already in the buffer. For the headers you need a loop to keep reading into the buffer until you reach the end. For the body, you don't want to eagerly read to the end, you want a stream that wraps the pipe and lazily reads to the end.

Reply:
Lets move this discussion to the PR itself and flesh out a design.

Reply:
I am not sure how to create a Stream for that using PipeReader -
The current implementation basically sends the entire stream, with a StartPosition for the stream and a Length - which is enough to easily read a part of a stream.

It seems that the only way to change a PipeReader's read cursor is by using AdvanceTo, which from the docs seems like it only moves forward. I will try things but your guidance would be appreciated.

Reply:
Look at the existing MultipartReaderStream and how it wraps the inner stream. You should be able to do about the same wrapping a pipe.

Reply:
I think I know how to make it work, however I did look into FormFile, and I don't know if it will break it. 

It understandably requires a BaseStreamOffset, because it seems to continue reading the stream and keep a position and a length for future processing - it's not something that can be easily achieved by PipeReader, but I think that by summarizing all the bytes consumed I can create a BaseStreamOffset, do you think it will work?

Reply:
By the time you create a FormFile each multipart section needs to be fully buffered. The old design used a single buffer stream with baseStreamOffset indicating the start of a section. When using pipes it might be easier to buffer the sections separately and pass each FormFile its own buffer stream with an offset of 0.